patchwork-extensions-blockentity/src/main/java/net/patchworkmc/mixin/extensions/blockentity/MixinWorld.java

```diff
@@ -0,0 +1,178 @@
+
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.util.Tickable;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.profiler.Profiler;
+import net.minecraft.world.World;
+import net.minecraft.world.chunk.WorldChunk;
+
+/**
+ * This Mixin implements {@link IForgeTileEntity#onLoad()} and {@link IForgeTileEntity#onChunkUnloaded()}.
+ */
+@Mixin(World.class)
+public class MixinWorld {
+	/////////////////////////////////////
+	/// addBlockEntity()
+	/////////////////////////////////////
+	@Inject(method = "addBlockEntity", at = @At("HEAD"))
+	private void onAddBlockEntity(BlockEntity blockEntity, CallbackInfoReturnable<Boolean> ci) {
+		World me = (World) (Object) this;
+
+		if (blockEntity.getWorld() != me) {
+			// Forge - set the world early as vanilla doesn't set it until next tick
+			blockEntity.setWorld(me);
+		}
+	}
+
+	@Shadow
+	@Final
+	protected List<BlockEntity> pendingBlockEntities;
+
+	@Inject(method = "addBlockEntity", cancellable = true,
+			at = @At(value = "INVOKE", ordinal = 0, shift = Shift.AFTER, remap = false,
+			target = "org/apache/logging/log4j/Logger.error(Ljava/lang/String;[Lorg/apache/logging/log4j/util/Supplier;)V"))
+	private void onBlockEntityAdding(BlockEntity blockEntity, CallbackInfoReturnable<Boolean> cir) {
+		// Forge: wait to add new TE if we're currently processing existing ones
+		cir.setReturnValue(pendingBlockEntities.add(blockEntity));
+	}
+
+	@Inject(method = "addBlockEntity", at = @At(value = "FIELD", ordinal = 0, shift = Shift.BEFORE,
+			target = "net/minecraft/world/World.isClient:Z", opcode = Opcodes.GETFIELD))
+	private void onBlockEntityAdded(BlockEntity blockEntity, CallbackInfoReturnable<Boolean> cir) {
+		// We cannot use Fabric events because Forge's onLoad() is also called on the client side.
+		((IForgeTileEntity) blockEntity).onLoad();
+	}
+
+	/////////////////////////////////////
+	/// tickBlockEntities()
+	/////////////////////////////////////
+	@Shadow
+	protected boolean iteratingTickingBlockEntities;
+
+	@Inject(method = "tickBlockEntities", at = @At(value = "INVOKE", ordinal = 0, shift = Shift.AFTER,
+			target = "net/minecraft/util/profiler/Profiler.push(Ljava/lang/String;)V"))
+	private void onTickBlockEntitiesStart(CallbackInfo ci) {
+		// Forge: Move above remove to prevent CocurrentModificationException
+		iteratingTickingBlockEntities = true;
+	}
+
+	@Shadow
+	@Final
+	protected List<BlockEntity> unloadedBlockEntities;
+
+	@Inject(method = "tickBlockEntities", at = @At(value = "INVOKE", ordinal = 0, shift = Shift.BEFORE,
+			target = "java/util/List.removeAll(Ljava/util/Collection;)Z"))
+	private void onBlockEntitiesRemoved(CallbackInfo ci) {
+		for (BlockEntity blockEntity: unloadedBlockEntities) {
+			((IForgeTileEntity) blockEntity).onChunkUnloaded();
+		}
+	}
+
+	@Unique
+	private static final ThreadLocal<BlockEntity> onBlockEntitiesRemoved_BlockEntity = ThreadLocal.withInitial(() -> null);
```

Comment:
I don't think this part of World is thread-safe. Maybe implement something like what I do in my BlockSnapshot PR?

Reply:
`tickBlockEntities` is supposed to be fired on either the main server thread or the main client thread?

![image](https://user-images.githubusercontent.com/4704366/89445081-078d7a00-d796-11ea-8c99-8f73f4d412be.png)